### PR TITLE
Refactor BugsnagEvent constructors

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

The constructors for `BugsnagEvent` have been split into chained calls to make it more explicit that they are parsing 3 different types of data. They now all call a simple constructor which sets all the required fields on the object, making it more obvious which fields are required and which are optional.

## Changeset

The `initWithKSReport` constructor now chains a call into the following constructors depending on what JSON it is passed:

- `initWithOOMData`
- `initWithKSCrashData`
- `initWithUserData`

These all call into the `initWithApp...` constructor that simply sets fields on the object. This ensures that the internal state of `BugsnagEvent` should be more consistent.

## Tests

Relied on existing test coverage, which was updated to use the new constructors where appropriate.

### CODE REVIEW

File: Source/BugsnagEvent.m [hunk 1]
```diff
@@ -664,14 +667,14 @@ - (void)setUser:(NSString *_Nullable)userId
...
-- (BugsnagUser *)parseUser:(NSDictionary *)metadata {
+- (BugsnagUser *)parseUser:(NSDictionary *)metadata deviceAppHash:(NSString *)deviceAppHash {
     NSMutableDictionary *user = [metadata[BSGKeyUser] mutableCopy];
     if (user == nil) {
         user = [NSMutableDictionary dictionary];
     }
 
-    if (!user[BSGKeyId] && self.device.id) { // if device id is null, don't set user id to default
-        BSGDictSetSafeObject(user, self.deviceAppHash, BSGKeyId);
+    if (!user[BSGKeyId]) {
```
Comment (tomlongridge) [inline Source/BugsnagEvent.m:676]: Is deviceAppHash now guaranteed non-null?

File: Source/BugsnagEvent.m [hunk 2]
```diff
@@ -304,189 +304,252 @@ - (BOOL)shouldBeSent;
...
+
+    BugsnagEvent *obj = [self initWithApp:[BugsnagAppWithState appFromJson:bugsnagPayload[@"app"]]
+                                   device:[BugsnagDeviceWithState deviceFromJson:bugsnagPayload[@"device"]]
+                             handledState:[BugsnagHandledState handledStateFromJson:bugsnagPayload]
+                                     user:[[BugsnagUser alloc] initWithDictionary:bugsnagPayload[@"user"]]
+                                 metadata:[[BugsnagMetadata alloc] initWithDictionary:bugsnagPayload[@"metaData"]]
+                              breadcrumbs:[BugsnagBreadcrumb breadcrumbArrayFromJson:bugsnagPayload[@"breadcrumbs"]]
+                                   errors:errors
+                                  threads:threads
+                                  session:nil];
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply Source/BugsnagEvent.m:548]: The session object isn't currently deserialized from the bugsnag payload for handled errors - there's an existing ticket to track this. The original intention was to address this issue separately to produce an easier changeset for review

File: Source/BugsnagEvent.m [hunk 3]
```diff
@@ -304,189 +304,252 @@ - (BOOL)shouldBeSent;
...
-            }
+/**
+ * Creates a BugsnagEvent from OutOfMemory JSON. OOMs contain less information
+ * than a regular crash report - notably they have no stacktrace or thread information.
+ *
+ * @param event a KSCrash report
+ *
+ * @return a BugsnagEvent containing the parsed information
+ */
+- (instancetype)initWithOOMData:(NSDictionary *)event {
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply Source/BugsnagEvent.m:379]: `errorType` is no longer required as a property from what as I can see - all the calculations involving `errorType` now take place in the constructor. The same is true for `error` - it has been superseded by the `errors` property.

I'll remove these fields entirely as they appear to be unused and likely to cause confusion.

File: Source/BugsnagEvent.m (hunk 1 above)
Comment (fractalwrench) [reply Source/BugsnagEvent.m:676]: `deviceAppHash` should be non-null - I think it'd be preferable to alter this to `BSGDictInsertIfNotNil`, and also pass in `self.device.id` if it's available to retain the old logic.

File: Source/BugsnagEvent.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply Source/BugsnagEvent.m:379]: The `error` property does appear to be used in the payload so I've assigned the value in this constructor by retrieving the value (if any) from the metadata.

**Pull request merged**